util/repr-bytes/src/lib.rs

```diff
@@ -0,0 +1,492 @@
+    fn from_bytes(src: &GenericArray<u8, Self::Size>) -> Result<Self, Self::Error>;
+
+    /// Convert to canonical representation bytes.
+    fn to_bytes(&self) -> GenericArray<u8, Self::Size>;
+
+    /// In-place visitation of the canonical bytes representation, using a closure
+    ///
+    /// Implementation note: The default implementation is not the best when your
+    /// type implements AsRef<[u8]>, it will make a needless copy in that case.
+    /// If your type implements AsRef<[u8]>, then you are strongly recommended to use
+    /// the macro `derive_repr_bytes_from_as_ref_try_from`.
+    /// Otherwise the default implementation is probably the best.
+    /// See also the suggested impl `derive_into_vec_from_repr_bytes`.
+    fn map_bytes<T, F: FnOnce(&[u8]) -> T>(&self, f: F) -> T {
+        f(self.to_bytes().as_slice())
+    }
+
+    /// Convenient helper: Get the representation size as a usize
+    fn size() -> usize {
+        Self::Size::USIZE
+    }
+}
+
+/// Error that indicates that we got a different number of bytes than expected
+#[derive(Debug)]
+pub struct LengthMismatch {
+    expected: usize,
+    found: usize,
+}
+
+impl Display for LengthMismatch {
+    fn fmt(&self, formatter: &mut core::fmt::Formatter) -> core::fmt::Result {
+        write!(
+            formatter,
+            "Expected exactly {} bytes, got {}",
+            self.expected, self.found
+        )
+    }
+}
+
+////
+// Suggested Implementations:
+// These macros provide instances of what can be thought of as "blanket implementations"
+// on a per-type basis.
+// These are macros because rust's coherence rules mean that blanket implementations
+// won't work out well for this use-case.
+//
+// There are two types of suggested impls:
+// - Impls of other traits in terms of ReprBytes
+// - Impls of ReprBytes in terms of other traits
+////
+
+/// Derive ReprBytes from AsRef<[u8]>, TryFrom<&[u8]>, and Size as a typenum.
+/// This is expected to be the right implementation for almost all cryptographic
+/// primitives, e.g. X25519, CompressedRistretto, etc.
+/// It can't work for e.g. RistrettoPoint, which doesn't have AsRef<[u8]>.
+///
+/// Arguments:
+///   - $mytype is the type you want to impl ReprBytes
+///   - $mysize is a typenum, representing the size of the canonical representation
+///
+/// Requirements:
+///   - <AsRef<[u8]> for $mytype>::as_ref().len() always equals $mysize::USIZE
+///   - <TryFrom<&[u8]> for $mytype>::Error implements core::fmt::Display
+///   - <TryFrom<&'a[u8]> for $mytype>::Error is the same for all values of 'a,
+///     OR they are all convertible to the value when 'a = 'static, via core::convert::From.
+#[macro_export]
+macro_rules! derive_repr_bytes_from_as_ref_and_try_from {
+    ($mytype:ty, $mysize:ty) => {
+        impl $crate::ReprBytes for $mytype {
+            type Size = $mysize;
+            type Error = <$mytype as TryFrom<&'static [u8]>>::Error;
```

Comment:
You should make `TryFrom` a re-export from this crate, and then reference it as `$crate::_exports::TryFrom` (or whatever).

Reply:
the user cannot choose not to import core, so i just went with `::core::convert::TryFrom`